tests/test_email.py

```diff
@@ -322,8 +355,123 @@ def test_decode_invitation_token_does_not_work_if_token_expired(email_app):
+        self.template_id = 'my-cool-example-template'
+        self.personalisation = OrderedDict([
+            ('foo', 'foo_p13n'),
+            ('bar', 'bar_p13n'),
+            ('baz', 'baz_p13n'),
+        ])
+        self.time = datetime.now()
+        self.test_api_key = '1111111111222222222233333333334444444444555555555566666666667777777777788888888888'
+        self.standard_args = {
+            'email_address': self.email_address,
+            'template_id': self.template_id,
+            'personalisation': self.personalisation
+        }
+
+    def test_get_reference(self, dm_notify_client):
+        """Test the staticmethod that creates the reference field to be sent with the request."""
+        ref = dm_notify_client.get_reference(**self.standard_args)
+        expected = (
+            '{"email_address": "example@example.com", "template_id": "my-cool-example-template", '
+            '"personalisation": {"foo": "foo_p13n", "bar": "bar_p13n", "baz": "baz_p13n"}, '
+            '"time": "2016-01-01T12:00:00.000000Z"}'
+        )
+
+        assert ref == expected
+
+    def test_decode_reference(self, dm_notify_client):
+        """Test the function that decodes the reference once it's returned from the api."""
+        decoded = dm_notify_client.decode_reference(dm_notify_client.get_reference(**self.standard_args))
+        expected = {
+            'personalisation': {
+                'foo': 'foo_p13n',
+                'bar': 'bar_p13n',
+                'baz': 'baz_p13n',
+            },
+            'email_address': 'example@example.com',
+            'template_id': 'my-cool-example-template',
+            'time': datetime.strptime('2016-01-01T12:00:00.000000Z', DATETIME_FORMAT)
+        }
+
+        assert decoded == expected
+
+    def test_send_email_creates_reference(self, dm_notify_client, dm_notify_example_success):
+        """Test the send_email function with data that should pass and run without exception."""
+        with mock.patch('notifications_python_client.NotificationsAPIClient.send_email_notification') as email_mock:
+            email_mock.return_value = dm_notify_example_success
+            dm_notify_client.send_email(self.email_address, self.template_id)
+
+            email_mock.assert_called_with(
+                self.email_address,
+                self.template_id,
+                personalisation=None,
+                reference=''.join([
+                    '{"email_address": "example@example.com", "template_id": "my-cool-example-template", ',
+                    '"personalisation": null, "time": "2016-01-01T12:00:00.000000Z"}'
+                ]),
+            )
+
+    def test_custom_reference(self, dm_notify_client, dm_notify_example_success):
+        """Test we can override the auto reference."""
+        with mock.patch('notifications_python_client.NotificationsAPIClient.send_email_notification') as email_mock:
+            dm_notify_example_success.update(reference='')
```

Comment:
I don't think we need to update the mock response at all here if we're not checking it anywhere?